src/sys/pthread.rs

```diff
@@ -34,10 +41,285 @@ pub fn pthread_kill<T>(thread: Pthread, signal: T) -> Result<()>
+/// let mutex = unsafe { &*mutex_ptr };
+/// 
+/// // If transmute or cast into a mutex, you must ensure it is initialized.
+/// // By default mutex's are process private, so we need to initialize with the `MutexAttr` with
+/// // shared.
+/// let mut mutex_attr = MutexAttr::new()?;
+/// mutex_attr.set_shared(true)?;
+/// mutex.init(Some(mutex_attr))?;
+/// 
+/// match unsafe { fork()? } {
+///     ForkResult::Parent { child } => {
+///         assert_eq!(mutex.lock(), Ok(()));
+///         sleep(TIMEOUT);
+///         assert_eq!(mutex.unlock(), Ok(()));
+///         // Wait for child process to exit
+///         unsafe {
+///             assert_eq!(libc::waitpid(child.as_raw(),std::ptr::null_mut(),0),child.as_raw());
+///         }
+///     },
+///     ForkResult::Child => {
+///         let now = Instant::now();
+///         sleep(DELTA);
+///         assert_eq!(mutex.lock(), Ok(()));
+///         assert!(now.elapsed() > TIMEOUT && now.elapsed() < TIMEOUT + DELTA);
+///     }
+/// }
+/// 
+/// # Ok(())
+/// # }
+/// ```
+#[cfg(target_os = "linux")]
+#[derive(Debug)]
+pub struct Mutex(UnsafeCell<libc::pthread_mutex_t>);
+#[cfg(target_os = "linux")]
+impl Mutex {
+    /// Wraps [`libc::pthread_mutex_init`].
+    pub fn init(&self, attr: Option<MutexAttr>) -> Result<()> {
+        let attr = match attr {
+            Some(mut x) => &mut x.0,
+            None => std::ptr::null_mut()
+        };
+        unsafe {
+            Errno::result(libc::pthread_mutex_init(self.0.get(),attr))?;
+        }
+        Ok(())
+    }
+    /// Wraps [`libc::pthread_mutex_init`].
+    pub fn new(attr: Option<MutexAttr>) -> Result<Self> {
+        let attr = match attr {
+            Some(mut x) => &mut x.0,
+            None => std::ptr::null_mut()
+        };
+        let init = unsafe {
+            let mut uninit = std::mem::MaybeUninit::<libc::pthread_mutex_t>::uninit();
+            Errno::result(libc::pthread_mutex_init(uninit.as_mut_ptr(),attr))?;
+            uninit.assume_init()
+        };
+        Ok(Self(UnsafeCell::new(init)))
+    }
+    /// Wraps [`libc::pthread_mutex_lock`].
+    /// 
+    /// <https://man7.org/linux/man-pages/man3/pthread_mutex_lock.3p.html>
+    pub fn lock(&self) -> Result<()> {
+        unsafe {
+            Errno::result(libc::pthread_mutex_lock(self.0.get())).map(drop)
+        }
+    }
+    /// Wraps [`libc::pthread_mutex_trylock`].
+    /// 
+    /// <https://man7.org/linux/man-pages/man3/pthread_mutex_lock.3p.html>
+    pub fn try_lock(&self) -> Result<bool> {
+        unsafe {
+            match Errno::result(libc::pthread_mutex_trylock(self.0.get())) {
+                Ok(_) => Ok(true),
+                Err(Errno::EBUSY) => Ok(false),
+                Err(err) => Err(err)
+            }
+            
+        }
+    }
+    /// Wraps [`libc::pthread_mutex_unlock`].
+    /// 
+    /// <https://man7.org/linux/man-pages/man3/pthread_mutex_lock.3p.html>
+    pub fn unlock(&self) -> Result<()> {
```

Comment:
According to https://man7.org/linux/man-pages/man3/pthread_mutex_lock.3p.html , unlocking a pthread mutex from a thread which does not own it is undefined behavior.  So this API isn't sound.  Perhaps you can do better if the lock function returns a Guard which is neither `Send` nor `Sync`, and you can only unlock the mutex via the Guard.

Reply:
Added a guard and made `Mutex::unlock` unsafe and documented with this.